### ISSUE TITLE
Feature request: Support for linear bacterial replicons?

### ISSUE THREAD

**Issue by tuspjo** [gbouras13/dnaapler#82]
Thank you for making and maintaining dnaapler! 

Streptomyces and other genera have linear chromosomes and often also linear plasmids.

Multiple genes on these replicons are located in a conserved fashion in relation to each other, why it would be practical to orient the replicons according to genes such as the dnaA gene on the chromosome, or replication genes on plasmids. 

If dnaapler could have a --linear flag to not rotate the replicons, but still orient them so the key gene is always on the positive strand, that would aid uniformity of assemblies. 

Best, 

Tue

**Comment by gbouras13** [comment]
Hi @tuspjo ,

To be clear, you'd simply want a functionality such that e.g. the dnaA gene would always be on the forward strand, but no actual reorientation (as they are linear?) - I guess that would just involve detecting where the dnaA gene is and reverse complimenting the contig if it is negative stranded. Seems straightforward.

George

**Comment by tuspjo** [comment]
Hi George, 
Yes, exactly, ensuring dnaA on the forward strand without rotating. 

Best, 

Tue.

**Comment by MrTomRod** [comment]
(I'm adding this here instead of creating a new issue, hope that's okay.)

I use [NCBI PGAP FASTA headers](https://github.com/ncbi/pgap/wiki/Input-Files#genome-assembly-sequence-file) after assembly and this may look as follows:

```
>XXX_scf1 [topology=circular] [location=chromosome]
ATGC...
>XXX_scf2 [topology=circular] [location=plasmid] [plasmid-name = pXXX-1]
ATGC...
>XXX_scf2 [topology=linear] [location=plasmid] [plasmid-name = pXXX-2]
ATGC...
```

For me, the ideal behaviour would be if Dnaapler could respect this notation.

I recommend the following check:

```python
header = '>XXX_scf2 [topology=linear] [location=plasmid]'
is_linear = '[topology=linear]' in '[  topology  = linear]'.replace(' ', '')  # NCBI notation allows for blanks
```

And this behaviour:
```python
if is_linear:
    reorient_only()
else:
    reorient_and_rotate()
```

**Comment by tuspjo** [comment]
As examples of why the linearity check is important, here are 9 refseq chromosomes marked linear but which I suspect are rotated to start with the dnaA gene (I guess the topology could also be wrongly assigned, but since _Streptomyces_ chromosomes are almost always linear, I think rotation is the issue.). 
NZ_CP046623.1
NZ_CP030862.1
NZ_CP047146.1
NZ_CP136652.1
NZ_CP169412.1
NZ_CP016279.1
NZ_CP029241.1
NZ_CP134875.1
NZ_CP016279.1

**Comment by gbouras13** [comment]
@tuspjo - very interesting. 

2 things to add:

1. I will aim to support this in the next version of dnaapler (with a more full refactor) - when that will be I can't promise.
2. The one and only @rrwick added and implemented .gfa support recently for dnaapler, that I have merged in - this should "solve" the problem for .gfa inputs at least, as only circular contigs will be rotated.

George

**Comment by tuspjo** [comment]
Ah, wonderful. And since rotation should be done before polishing, not after, this solution would solve a bunch of things.

**Comment by tuspjo** [comment]
> Btw, I noticed that the files produced by Flye sometimes disagree: [mikolmogorov/Flye#659](https://github.com/mikolmogorov/Flye/issues/659)
yes, as I understand it GFAs are kind of tmp files and pre polishing. Since I have been using a custom GFA-input contigger, I have used the flye polishing standalone after contigging so the workflow is Flye->GFAcontigging->Flyepolish->polypolish->POLCA. But I agree that some could miss the difference and just use the GFA as if it was a polished genome. So allowing (flye)GFA input comes with the risk that people might think they're rotating a polished genome when they're not.

**Comment by mbhall88** [comment]
So in order to address all the valid suggestions in this issue I think there is a couple of things that could be implemented:
- From GFA: Ensuring linear sequences that contain a "replication" gene (i.e., dnaA, repA etc.) have said gene on the forward strand
- From FASTA: Detect some header annotations that we are clear about upfront—the NCBI standard `topology=linear` seems best—and ensuring the detected replication gene is on the forward strand. 

After thinking about this issue a lot this morning it definitely feels like GFA is the best input as we don't really need to infer much about the topology as it is all contained in the assembly graph. Relying on FASTA headers isn't too bad though I guess.

Thoughts?

**Comment by tuspjo** [comment]
GFA input will have base quality compromises as it's pre-polishing. 
For Streptomyces, 9/10 times the genome is not completely resolved in the GFA. But if semicomplex graphs like the ones in the inserted image could be easily deciphered as linear I also see the benefits of directly using the graph format and not relying on a text string. 

<img width="1220" height="890" alt="Image" src="https://github.com/user-attachments/assets/1eae6201-106d-4146-b0f9-3606bf4c0a14" />

**Comment by mbhall88** [comment]
I guess my thinking was the GFA of a resolved assembly

**Comment by tuspjo** [comment]
Ah yes, ok. For other genera that would likely work, but unless stringency is relaxed from the assemblers, Streptomyces genomes are unlikely to resolve based on reads reaching from the core to the telomeric end as this distance is often 100k-1M nt. 

<img width="1109" height="813" alt="Image" src="https://github.com/user-attachments/assets/53dbd044-bfdf-4c58-8a32-89025cc1dd87" />
